ti/instrumentation/utils.go

```diff
@@ -264,6 +266,136 @@ func getChangedFiles(ctx context.Context, workspace string, log *logrus.Logger,
+				moduleName := splitName[0]
+				moduleList = append(moduleList, moduleName)
+			}
+		} else {
+			//get list of .java files from the output
+			javaFileNames := extractJavaFilesFromQueryOutput(allJavaSrcsOutput)
+			for _, name := range javaFileNames {
+				//to prevent duplicate files
+				if _, exists := uniqueFiles[name]; !exists {
+					changedRes = append(changedRes, ti.File{Status: file.Status, Name: name})
+					uniqueFiles[name] = struct{}{}
+				}
+			}
+		}
+	}
+	return changedRes, moduleList, nil
+}
+
+// takes bazelQueryOutput string and parses it to extract .java files and returns in a list
+func extractJavaFilesFromQueryOutput(bazelOutput []byte) []string {
+	var javaFileNames []string
+	// Outer Regex to extract only srcs
+	pattern1 := `srcs = \[[^\]]+\]`
+	r1 := regexp.MustCompile(pattern1)
+
+	sections := r1.FindAllString(string(bazelOutput), -1)
+	for _, section := range sections {
+		// Inner Regex to match only .java files inside each src
+		pattern2 := `"//[^"]+\.java"`
+		r2 := regexp.MustCompile(pattern2)
+		matches := r2.FindAllString(section, -1)
+		for _, match := range matches {
+			srcFile := strings.Split(strings.Trim(match, `\"`), ":")
+			changedFile := strings.TrimLeft(string(srcFile[0]), "//") + "/" + string(srcFile[1])
+			javaFileNames = append(javaFileNames, changedFile)
+		}
+	}
+	return javaFileNames
+}
+
+// takes file name, runs bazel queries extract src globs defined in java rules, viz java_library and java_test
+func getJavaRulesFromBazel(file string, workspace string, ctx context.Context) ([]byte, []byte, []byte, error) {
+	bazelQueryInput := strings.Replace(file, "/BUILD.bazel", "", -1)
+	//query to fetch test src globs from java rule
+	c1 := fmt.Sprintf("cd %s; %s query 'kind(\"java\", %s:*)' --output=build | grep 'srcs =' ", workspace, bazelCmd, bazelQueryInput)
+	cmdArgs1 := []string{"-c", c1}
+	javaKindOutput, err := execCmdCtx(ctx, "sh", cmdArgs1...).Output()
+	if err != nil {
+		fmt.Errorf("failed to run bazel query %v, encountered %v as %v has no java rule", c1, err, file)
+	}
+	//query to count the scr files in bazelQueryOutput
+	c2 := fmt.Sprintf("cd %s; %s query 'kind(\"java\", %s:*)' --output=build | grep 'srcs =' | grep -o '\\.java' | wc -l", workspace, bazelCmd, bazelQueryInput)
+	cmdArgs2 := []string{"-c", c2}
+	countQueryOutput, err := execCmdCtx(ctx, "sh", cmdArgs2...).Output()
+
+	//query to get test kind for module list
+	c3 := fmt.Sprintf("cd %s; %s query 'kind(\"java_library\", %s:tests)' --output=build | grep 'srcs =' ", workspace, bazelCmd, bazelQueryInput)
+	cmdArgs3 := []string{"-c", c3}
+	javaTestKindOutput, err := execCmdCtx(ctx, "sh", cmdArgs3...).Output()
+
+	return javaKindOutput, countQueryOutput, javaTestKindOutput, err
+}
+
+// takes a directory name and adds all java files within that module/package in changed file res
+func getAllJavaFilesInsideDirectory(directory string, res []ti.File, file ti.File, uniqueFiles map[string]struct{}) ([]ti.File, error) {
```

Comment:
I believe we already have a util for this.

Reply:
do you mean Go utility? It has but I want to just consider .java files not scripts and other files.

Reply:
Please name the params to indicate their purpose. `res` in the input doesn't sound right.